### ISSUE TITLE
Proposal preferences

### ISSUE THREAD

**Issue by wendorf** [cloudfoundry/cc-api-v3-style-guide#11]
Please add your proposal preferences to this issue. Mine are as follows:

Status codes:
I prefer proposal 3 (include error code, not HTTP status code). Status codes are available in the response object and not super helpful in the body. Error codes are very easily searchable, thus more useful.

Relationships:
I prefer proposal A since it provides a very clear way of managing relationships. A1 seems overly verbose and will be cumbersome to CC consumers who are not interested in relationships.

Including Related Resources:
Proposal B is too wordy. If we wanted to include "spaces" by default in an org, we would do my_org.spaces, not my_org.included.spaces. Proposal A is much better. Proposal C is too confusing.

Asynchronicity:
Get rid of the async flag. App staging? Always asynchronous. /v3/info? Always synchronous. This puts us in hot water if we have a synchronous endpoint that needs to be changed to asynchronous later during the life of V3. Any ideas for what we can do there?

**Comment by zrob** [comment]
Status codes:
Proposal A.  I don't think having the status in the body is very valuable.  Having the arbitrary code may be useful, but I don't think we need it right now and it could be an additive change later.

Relationships:
I perfer proposal A as it seems less verbose and more clear.

Related Resources:
Can't decide yet.  Proposal A is simple. Proposal B distinguishes between real fields and abnormal extra thingies.  Proposal C allows de-duping and a flatter response.

**Comment by utako** [comment]
Status Code: I like Proposal C. I agree with Dan; I appreciate its searchable quality. Might provide a good opportunity to improve our logging. I don't find much value in Proposal B.

Relationships: Ditto what @wendorf and @zrob said about A1. 

Related Resources: Proposal B is explicit about what is being requested, but as a consumer of the API, I may find the nested "included" key tedious and it would probably sadden me. Proposal A is simple and easy to parse, but may be susceptible to key chain duping?

**Comment by ships** [comment]
Status codes: 3. Error code has been useful to me at least once, they are easy to find and read a little about. However, response code is in the header already, no need for duplication.

Relationships: A. simpler, less redundant.

Related resources: I like A more than the other two, though C has a nice parallel in the Relationships thing from above. Rejects B. Related; I like the filtering subresource results by field.

Asynchronicity: This proposal is interesting, but i don't quite buy it as it stands. What if we follow via @wendorf's suggestion above, and remove the async flag-- such that `GET` requests are always synchronous, whereas `PUT, POST, PATCH, DELETE` are all asynchronous; then the response to the async requests will include the current status of the thing being operated on (which includes a "self" link, in case we want to keep asking for this info or watch for the change to finish) as well as the job url as specified here -- GETing on the job url should return the same response as this initial one?

I suspect it would be reasonable to make all data-changing endpoints asynchronous, because 

example:

```
PUT /v3/apps/:guid -d {"some_field":"value"}

202 Accepted
{
    "job_url":"/v3/jobs/123",
    "job_status":"pending",
    "resource": {
        "guid": ...
        "some_field":"old_value",
        "_links": {
            "self":"/v3/apps/:guid"
        }
    }
}
```

sometime later, possibly right away ish

```
GET /v3/jobs/123

200 OK
{ ...
    "job_status":"done"
 ...
    "resource": {
        "some_field":"value"
...
}
```

**Comment by Gerg** [comment]
From first cage match:  
For relationships proposal: People like A.  
Jury is still out for including related resources. Slight preference for A?

**Comment by Gerg** [comment]
From second cage match:

For related resources: C, assuming we can figure out the glue.
